### ISSUE TITLE
Switchable default endianness

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#17]
There are quite a few binary formats that allow 2 versions: little-endian and big-endian and have some sort of indicator near the beginning of the file to determine which version we're reading now. 2 "versions" differ only in one aspect: default endianness.

Examples of such format:
- [TIFF](http://www.fileformat.info/format/tiff/corion.htm): it starts with 2 bytes, `II` for l-e and `MM` for b-e version
- [pcap](https://wiki.wireshark.org/Development/LibpcapFileFormat): it starts with 4 bytes, `[0xd4, 0xc3, 0xb2, 0xa1]` for l-e and `[0xa1, 0xb2, 0xc3, 0xd4]` for b-e
- [elf](http://geezer.osdevbrasil.net/osd/exec/elf.txt): byte 5 determines endianness of all integers in the file

Such formats are definitely readable with KS right now, but it means duplicating whole format description twice - once with a `endian: le` and once with `endian: be` and then switching between them with `if: ...`.

There's a [proposal](https://www.linux.org.ru/news/opensource/12797115?lastmod=1470832752452#comment-12798628) to introduce a concept of switchable default endianness:

``` yaml
meta:
  id: tiff
seq:
  - id: byte_order_id
    type: str
    encoding: ASCII
    size: 2
  - id: header
    type: tiff_header
types:
  tiff_header:
    meta:
      little-endian-expr: '_root.byte_order_id == "II"'
    seq:
      - id: version
        type: u2
      - id: img_dir_ofs
        type: u4
      # ...
```

Basically, it means for `tiff_header` type and all types defined underneath it should determine its endianness by the result of evaluation of expression given in `little-endian-expr`.

Things I'd like to discuss:
- I don't like the name of `little-endian-expr`
- Implementation of these ones

If we want to aim for maximum performance, probably that means that compiler should generate two distinct versions of the classes like `tiff_header` (and everything beneath that in hierarchy) and choose which one they would use once. This would result in generation of classes like `_be_TiffHeader` and `_le_TiffHeader` - and, for statically typed languages, probably it would be a good idea to generate a common interface `TiffHeader` for them.

**Comment by LogicAndTrick** [comment]
Some random thoughts, but I don't think they're particularly good after thinking about them:

Maybe use the same syntax as the `repeat` option and have something like:

```yaml
types:
  tiff_header:
    meta:
      endian: expr
      endian-expr: '_root.byte_order_id == "II" ? "le" : "be"'
```

Though it doesn't look very nice to use a ternary expression for something that could be a boolean value. Maybe two options for flexibility:

```yaml
types:
  tiff_header:
    meta:
      endian: expr
      # only one of these would actually be used:
      endian-expr-le: '_root.byte_order_id == "II"' # LE if true
      endian-expr-be: '_root.byte_order_id != "II"' # BE if true
```

Usually you'd want to set the global default at runtime so it'd be nice if you could do it once and not need to specify it in every type:

```yaml
meta:
  endian: expr
  endian-expr-le: '_root.byte_order_id == "II"'   # lazy evaluated only when required
seq:
  - id: byte_order_id
    type: str
    encoding: ASCII
    size: 2
  - id: header
    type: tiff_header
types:
  tiff_header:
    seq:
      - id: version
        type: u2 # endian-expr is evaluated now
      # ...
```

As for generated classes, potentially you could generate code like this:

```cs
public TiffHeader(...)
{
    if (_root.byte_order_id == "II") {
        _parse_le();
    } else {
        _parse_be()
    }
}

private void _parse_be() { ... }
private void _parse_le() { ... }
```

Which would avoid the need to add more classes, and shouldn't have any significant performance impacts. You'd need to do the same thing in instances too though, so having multiple types is probably the easier approach.

**Comment by GreyCat** [comment]
> Maybe use the same syntax as the repeat option and have something like:

I like your train of thought - at the very least, it's much more uniform than my proposal.

There's definitely some connotation with `repeat` + `repeat-expr` here. Both are switches for various modes, sometimes with an extra expression as a parameter. Also, `process` syntax is close - it's also a big switch with some optional expression(s)-as-a-parameter. `size` vs `size-eos` is not that far away too.

I'm pondering the idea to change them all into a more uniform construct that would give us something like:

* `repeat: eos`
* `repeat: expr(some_var + 42)`
* `process: zlib`
* `process: xor([some_var1 & 0xff, some_var2])`
* `endian: le`
* `endian: expr(_root.byte_order_id == "II")`
* `size: eos`
* `size: expr(5)`
* `size: expr(some_var + 42)`

Although `size: expr(...)` is definitely gross.

> Usually you'd want to set the global default at runtime so it'd be nice if you could do it once and not need to specify it in every type:

An interesting idea, but I'm not sure how to (1) implement it, (2) make some compile-time guarantees that the code actually works - i.e. endianness mode won't be accessed until the relevant identifier bytes are read.

> ```java
private void _parse_be() { ... }
private void _parse_le() { ... }
```

Also an interesting idea. Actually you can even propagate calculate endianness somehow to inner classes, i.e.:

```java
public class TiffInner {
  ...
  private void _parse_be() {
    ...
    someInner = new TiffInner(..., Endianness.BIG);
  }

  public static class TiffInner {
    public TiffInner(..., Endianness e) {
      if (e == Endianness.BIG) {
          _parse_be();
      } else {
          _parse_le();
      }
    }
    private void _parse_be() { ... }
    private void _parse_le() { ... }
  }
}
```

It's probably less of a hassle than creating 2 separate classes + interface for them.

**Comment by LogicAndTrick** [comment]
I like the idea of making the various different mode switches more consistent and extensible.

Just while you're thinking about possible syntax changes, I was going to make a proposal at some point for something like this to repeat until a certain condition is met:
```yaml
seq:
  - id: number
    type: u1
    repeat: until
    repeat-until: value == 0 # where `value` is the last value that was read
```

Hopefully something like that can be possible with whatever the new syntax might be.

---

Extending on the generated class code, I like the idea of passing the endianness into the constructor. You'd want to save it somewhere in the class so it can be used in instances:

```java
  public static class TiffInner {
    public TiffInner(..., Endianness e) {
      this._endianness = e;                            // <--- here
      if (e == Endianness.BIG) {
          _parse_be();
      } else {
          _parse_le();
      }
    }
    private void _parse_be() { ... }
    private void _parse_le() { ... }

    public Whatever getInstance() {
      if (_instance != null) return _instance;

      if (this._endianness == Endianness.BIG) {             // <--- and here
        _instance = _io.readU4be();
      } else {
        _instance = _io.readU4le();
      }

      return _instance;
    }
  }
```

**Comment by GreyCat** [comment]
Oh bummer. I totally forgot about the instances, you're right. I'll try to think on pros and cons of all the options we have (i.e. separate classes + interface vs two parsing methods + switching inside instances) and summarize it then.

**Comment by koczkatamas** [comment]
I would make endianness switch as transparent to the parser code and the user as it can be. I would simply call setEndianness on KaitaiStream and then _io.readU4 would automatically use readU4be or readU4le internally.

As the endianness can change multiple times and we have to restore the original endianness sometimes, I would make a restoreEndianness() method too which must be called in parallel with setEndianness after we go out of the 'endianness scope' and the KaitaiStream would store the endianness history in a stack structure.

Instance streams would inherit the endianness from the parent.

**Comment by davidhicks** [comment]
ESRI Shapefiles (.shp) contain fields which are big-endian, and other fields which are little-endian. Allowing "endian:" support within "seq:" would resolve this issue and allow Shapefiles to be parsed.

For reference, the ESRI Shapefile specification is available at: https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf

**Comment by GreyCat** [comment]
@davidhicks As far as I see in the specification, this format uses well-defined endianness for every particular structure it defines:

> The integers and double-precision integers that make up the data description fields in the
> file header (identified below) and record contents in the main file are in little endian (PC
> or Intel®) byte order. The integers and double-precision floating point numbers that make
> up the rest of the file and file management are in big endian (Sun® or Motorola®) byte
> order

This can be perfectly done right now, either by specifying `meta/endian` in this particular record, or by using explicit types that specify endianness (i.e. `u4le`, `f8be`, etc).

**Comment by davidhicks** [comment]
@GreyCat : thank you, I had missed that section of the user guide on appending "le" and "be" to the integer types to specify the endian. It looks like only integers are supported using this format, not floats? For example, f8le or f8be?

**Comment by GreyCat** [comment]
@davidhicks `f8le` / `f8be` work as well, and both `f4` and `f8` without explicit endianness draw their endianness from nearest `meta/endian` definition (i.e. as defined in current type, or, if it is not defined in current type, as defined in a type higher by nesting hierarchy, etc).

**Comment by luser** [comment]
I was reading the Katai Struct docs and had this exact question and stumbled upon this issue. I just want to mention two file formats I've written parsers for in the past that have a slightly different way of handling endianness. Both Mach-O files and minidump files can be written in either endianness. They each start with a 4-byte magic number, and when reading a file you read that and determine either "the file is in native endianness" "the file is in the opposite endianness" or "this is not a valid file".

Mach-O files start with `MH_MAGIC`, which is `0xfeedface`, and the headers also define `MH_CIGAM` as `0xcdfaedfe`. (Likewise there are a `MH_MAGIC_64` and `MH_CIGAM_64` for 64-bit Mach-O files.) "Fat" Mach-O binaries containing multiple architectures have their own fat header which starts with `FAT_MAGIC` (or `FAT_CIGAM`), and then contains a set of single-architecture Mach-O binaries, each of which starts with its own magic number (and can have its own endianness). Refs: [1](http://tfpwn.com/blog/mach-o-file-format.html) [2](https://lowlevelbits.org/parsing-mach-o-files/).

Minidump files start with `MINIDUMP_SIGNATURE`, which is `0x504d444d ('PMDM')`, and similarly parsers will check against that or the bitswapped version to determine whether the file matches the current endianness.

Given those examples, I think it would be useful to have a way to indicate not a literal endianness, but a derived "endianness matches the host" or "endianness is opposite of host" value that can come from the value of a magic number.

**Comment by GreyCat** [comment]
This is basically the same technique as everyone else uses. If we'll see from raw file perspective, for Mach-O, having `CD FA ED FE` means little-endian and `FE ED FA CE` means big-endian file, that's all. The same for Minidump.

Relating that to individual host is several layers lower than that, in some implementation. There would be implementations that benefit from having "matching the host" or "not matching the host", there would be implementations that don't (or at least we'll have no direct control on if this brings any benefits or not).

**Comment by GreyCat** [comment]
Ok, many months late, I'd like to present a proposed solution. I've pushed everything I've done on this subject up to date in [calc-endian branch of compiler](https://github.com/kaitai-io/kaitai_struct_compiler/tree/calc-endian) and [endian_expr_is branch of tests](https://github.com/kaitai-io/kaitai_struct_tests/tree/endian_expr_is).

Proposed syntax resembles switch (I've also tried other syntax variants, but ultimately this one proved to be most flexible and usable):

```yaml
meta:
  endian:
    switch-on: _root.indicator
    cases:
      0xfeedface: be
      0xcdfaedfe: le
seq:
  - id: some_int
    type: u4 # => this become either little or big endian
```

It works, and it is relatively easy to understand and use. As usual, one can use `_` as "else" case, or, if you want to be strict, just specify exact values, and parsing will break with exception if nothing will match (as it is impossible to run parsing without known endianness). All subtypes would inherit `meta` setting properly, so you really typically need only to include this once, near the top of the file.

## A word of caution

Note that endianess decision is done before any `seq` parsing, so it's impossible to do stuff like:

```yaml
meta:
  id: foo
  endian:
    switch-on: indicator
    cases:
      1: be
      2: le
seq:
  - id: indicator
    type: u1
  - id: some_int
    type: u4
```

One needs to clearly separate this into 2 types, one with `indicator` value and another one with everything else that is endianness-dependent.

## Gory details

Under the hood, however, everything is somewhat complex and took lots of time and effort. Generally, the idea is that we generate two `_read_be` and `_read_le` methods instead of just one `_read`, i.e. we generate something like that:

```ruby
        case _parent.indicator
        when 0xfeedface
          @_is_le = false
        when 0xcdfaedfe
          @_is_le = true
        end

        if @_is_le == true
          _read_le
        elsif @_is_le == false
          _read_be
        else
          raise Kaitai::Struct::Stream::UndecidedEndiannessError
        end
```

Two very similar (but distinct in enforced endianness) `_read` methods are thus generated for `seq` parsing. However, that's not all:

* Extra effort needed to properly parse `instances`. For that, we save calculated endianness decision in `_is_le` variable in the object, and analyze it during instance parsing, i.e.:

```ruby
      def inst_int
        return @inst_int unless @inst_int.nil?
        _pos = @_io.pos
        @_io.seek(2)
        if @_is_le
          @inst_int = @_io.read_u4le
        else
          @inst_int = @_io.read_u4be
        end
        @_io.seek(_pos)
        @inst_int
      end
```

* Calculated `meta/endian` inheritance is complex. Using `_root.foo` is easy (one can just re-evaluate this expression in every object, although that would be slow, of course), but using `_parent.foo` is totally impossible without special precautions: `_parent` would point to not what it was supposed to point to in children.

For that, I'm introducing special `InheritedEndian` value, and, if class uses "inherited calculated endianness", it will get extra endianness argument in the constructor, i.e.:

```ruby
        def initialize(_io, _parent = nil, _root = self, _is_le = nil)
          super(_io, _parent, _root)
          @_is_le = _is_le

          if @_is_le == true
            _read_le
          elsif @_is_le == false
            _read_be
          else
            raise Kaitai::Struct::Stream::UndecidedEndiannessError
          end
        end
```

and all calls to it would pass current `_is_le` to it, i.e.:

```ruby
@inst_sub = SubMainObj.new(@_io, self, @_root, @_is_le)
#                                              ^^^^^^^
```

For now, compilers for Ruby, Python, Java work with this syntax, and I plan to complete all others soon.

**Comment by davidhicks** [comment]
Will it be possible to set different endianess for different types defined within the same file? There are a number of formats I'm aware of which embed other formats within, but not encoded/compressed as would be the case with an archive. I suppose it could be possible some of those formats have a difference in endianess between the outermost "parent" file, and the inner embedded files?

**Comment by GreyCat** [comment]
@davidhicks Sure, every type can have it's calculated endianness. It's correctly inherited and, if necessary, it can be overidden at every level.

**Comment by GreyCat** [comment]
Everything seems to be done (except for Go, but it is still many miles away from serious parsing anyway). Please take a look and, if everything if ok, I'll merge these branches into master in several days.

**Comment by GreyCat** [comment]
I've done a [working implementation of ELF](https://github.com/kaitai-io/kaitai_struct_formats/blob/master/executable/elf.ksy) that supports both big- and little-endian ELFs (caught a little bug in process), so it kind of proves that it's working and all that stuff.

**Comment by GreyCat** [comment]
I've merged everything up to master and fixed remaining bugs. C# is somehow broken yet again, everything else seems to be more or less fine. I'm going to consider this solved, unless somebody objects.

**Comment by GreyCat** [comment]
A [recent commit](https://github.com/kaitai-io/kaitai_struct_compiler/pull/93) by @tecknicaltom revealed that current state of affairs is really broken, especially for debug mode.

I'll try to start from a specification of what we want to get. This is kind of a 3D table:

* "non-debug" vs "debug"
* "fixed endian" vs "calculated endian"
* various parts of generated code / user app
  * API call to do parsing
  * typical structure of a constructor(s)
  * typical structure of `_read` method
  * typical structure of `_read_be` / `_read_le` method

## Non-debug

### Fixed endian

* API call: `a = Foo.new(...)`
* constructor(s): have `_read` invocation in the end
* `_read` method: contains all the eager parsing code
* `_read_be` / `_read_le` method: none

### Calculated endian

* API call: `a = Foo.new(...)`
* constructor(s): have `_read` invocation in the end
* `_read` method: calculates endianness, memorizes it for instances, calls `_read_be` or `_read_le`
* `_read_be` / `_read_le` method: two copies of eager parsing code with differently forced endianness

## Debug

### Fixed endian

* API call: `a = Foo.new(...); a._read`
* constructor(s): no automatic `_read` call
* `_read` method: contains all the eager parsing code
* `_read_be` / `_read_le` method: none

### Calculated endian

* API call: `a = Foo.new(...); a._read`
* constructor(s): no automatic `_read` call
* `_read` method: calculates endianness, memorizes it for instances, calls `_read_be` or `_read_le`
* `_read_be` / `_read_le` method: two copies of eager parsing code with differently forced endianness

**Comment by GreyCat** [comment]
Ok, I've implemented this scheme. @tecknicaltom — please take a look at https://github.com/kaitai-io/kaitai_struct_compiler/commit/5e4dbb4d02b552de3af36addb1bfa95245b8e580. Off to CI it goes.

**Comment by arekbulski** [comment]
This seems to be the last item on 0.8 milestone, and its almost a year old. What is the status of this feature?

**Comment by GreyCat** [comment]
It is technically working, but I'm not very happy with what we've got here. And, anyway, before v0.8 some formats that might benefit from this feature (i.e. [elf](https://github.com/kaitai-io/kaitai_struct_formats/blob/master/executable/elf.ksy), [exif](https://github.com/kaitai-io/kaitai_struct_formats/blob/master/image/exif.ksy), [dicom](https://github.com/kaitai-io/kaitai_struct_formats/blob/master/image/dicom.ksy), etc).

The whole v0.8 is terribly behind the schedule, but I still need to find a day or two to sort all this minor things out and get it ready to ship finally...

**Comment by arekbulski** [comment]
Can I help you with it somehow?

**Comment by GreyCat** [comment]
If you mean #17 — sure, take a look at it, and tell me what you think of it. May be try porting some of the formats in the repository and see if you think it's usable in real life?

If you mean whole v0.8 — generally, we need to finalize a checklist and finish it. We'll need to publish lots of packages (i.e. all the packages for ksc for different OSes, ksv, runtimes etc), update the website, post the news on our site, push public anouncements on major news sites, etc, etc.

**Comment by arekbulski** [comment]
Do we keep this feature the way it is and upgrade later, or withhold shipping 0.8?

**Comment by GreyCat** [comment]
I believe we'll stay with this implementation at least for v0.8, and then we'll see. Closing this one and considering it complete at this stage.
